### DISCUSSION TITLE
Vike v1 server/client routing clarification wanted

### DISCUSSION THREAD

**Discussion by igolkotek** [vikejs/vike#1316] [Discussion]
Hi everyone, thank you for the great plugin.

I ask you please to help me to understand the client-server interaction in Vite env.

In general I try to make a Vite+Vike+Svelte app without Sveltekit.
The app uses SSR, my server sends the code to a browser, it passes hydration stage, and then I have a PWA with a server required for api + module + statics downloading only. So a browser takes care of routing.

As I understand my +config.h.js has to be like this:
```js
export default {
  clientRouting: true,
  hydrationCanBeAborted: true,
  meta: {
    onBeforeRender: {
      env: { server: true, client: true }
    }
  }
}
```
Client routing works like a charm with `vite dev`, **but if you need** `renderPage` and a custom server, the client routing breaks down and makes requests to server on every route.

Why?

Thanks!

**Comment by lourot** [comment]
> **but if you need** `renderPage` and a custom server

Hi @igolkotek, I don't understand this phrase. Could you elaborate and/or push a minimal example somewhere? Thanks!

**Reply by igolkotek** [reply]
Thank you for your response, @AurelienLourot . I'll try )

After some digging I found that the problem is deeper.

### Settings
I created 2 apps with the official templates vike-app for react and vue and tested them.
Nothing in the official code was changed but 2 settings below:

1. Added client routing to `+config.h.js` :
```
  clientRouting: true
```

2. Added a logger middleware to `server/index.js`:
```
function logger(req, res, next) {
  console.log("============= SERVER. request url:", new Date(), "\n", req.url)
  next()
}
```

3. I have to use `Nginx` as a server proxy.
It's configuration:
```
server {
	listen 80 default_server;
	server_name sitename;

	location / {
            include proxy_params;
	    proxy_pass http://localhost:5173/;
	}
}
```
### Testing

#### Browsing `localhost:5173`
Both dev and production servers work OK. Client routing is OK.


#### Browsing 'http://sitename' with reverse proxy

1. To be mentioned first that counters do not work, both in vue and react. At least after a while react counter started to count after moving to `about` page and back. Maybe server just needed some time,
 I do not care this, at least not now.
 
2. Client routing doesn't work with proxy, pls see the 1st 2 blue marks at the screenshot below.
3. Server hangs and browser's waiting for a response (about 1.5 minutes in my case - look at the time marks in the blue circle)
![server error](https://github.com/vikejs/vike/assets/1233562/d5aa382e-4815-43d3-a4a8-ce399bb22315)

4. These are those chunks a browser was waiting for during a 1.5 minutes:
![server_error_waiting_end](https://github.com/vikejs/vike/assets/1233562/5df75980-7fd3-4625-88c4-62958eaddf77)

Server hanging is unpredictable, but happens.

Thus I have to formulate my problems as:
1. With a reverse proxy client routing does not work
2. With a reverse proxy server hangs

I have found [description of server hanging](https://github.com/vikejs/vike/issues/1094) in the issues, but is was not solved.

Maybe it's better to move this question to the issues?

Thank you for reading to the end )

**Reply by lourot** [reply]
I see, I'm not sure if this is related to Vike if everything works without a server proxy? I'll see tomorrow if I can reproduce locally

**Reply by igolkotek** [reply]
OK, thank you.
I use this same reverse proxy for several other projects, including Sveltekit, and never have seen this issue.

**Reply by igolkotek** [reply]
Just some screenshot in server development mode.

![Screenshot from 2023-12-07 12-33-40](https://github.com/vikejs/vike/assets/1233562/bb99148b-ca7c-4c96-b9d7-cfea7f267e0d)

Response time was 1.3 min.

**Reply by igolkotek** [reply]
In production mode the result is the same but Network inspector shows nothing.

This is a pending step:
![Screenshot from 2023-12-07 12-39-37](https://github.com/vikejs/vike/assets/1233562/3529b252-6dbb-4cd2-a011-5c214adaa803)

The final display:
![Screenshot from 2023-12-07 12-41-17](https://github.com/vikejs/vike/assets/1233562/a90f388c-a335-4a9c-83f1-893481a7e8f0)

Details:
![Screenshot from 2023-12-07 12-41-37](https://github.com/vikejs/vike/assets/1233562/f4c49816-a465-45dd-b7f6-771064ac5470)

**Reply by lourot** [reply]
@igolkotek Can you please give me a full, yet minimal `nginx.conf` to try this out on my machine? With what you gave me I get

```
"server" directive is not allowed here in /etc/nginx/nginx.conf:1
```

then if I wrap it inside `http { ... }` I get `no "events" section in configuration`.

It's important you give me a reproducer I can very quickly spin up, without having to wrap my head around nginx (haven't touched this in a long time). Can you also please push the code of your Vike project somewhere, so I can just clone it? Thank you

**Reply by igolkotek** [reply]
default minimal `nginx.conf`

```
user       www www;  ## Default: nobody
worker_processes  5;  ## Default: 1
error_log  logs/error.log;
pid        logs/nginx.pid;
worker_rlimit_nofile 8192;

events {
  worker_connections  4096;  ## Default: 1024
}

http {
  include    conf/mime.types;
  include    /etc/nginx/proxy.conf;
  include    /etc/nginx/fastcgi.conf;
  index    index.html index.htm index.php;

  default_type application/octet-stream;
  log_format   main '$remote_addr - $remote_user [$time_local]  $status '
    '"$request" $body_bytes_sent "$http_referer" '
    '"$http_user_agent" "$http_x_forwarded_for"';
  access_log   logs/access.log  main;
  sendfile     on;
  tcp_nopush   on;
  server_names_hash_bucket_size 128; # this seems to be required for some vhosts

server {
	listen 80 default_server;
	server_name sitename;       # in /etc/hosts file you have to add the line: 127.0.0.1 sitename
	location / {
	    proxy_pass http://localhost:5173/;
	}
}
}

```

**Reply by igolkotek** [reply]
I named a proxy server `sitename`.
Please add  the line in `/etc/hosts` file: 
```
127.0.0.1 sitename
```

**Reply by igolkotek** [reply]
@AurelienLourot 
"Can you please give me a full, yet minimal nginx.conf to try this out on my machine? With what you gave me I get"

Nginx installation generates default `nginx.conf` file.

You could just append my lines
```
server {
	listen 80 default_server;
	server_name sitename;

	location / {
	    proxy_pass http://localhost:5173/;
	}
}
```
into the end of `nginx.conf` http {} section.

**Reply by lourot** [reply]
> into the end of `nginx.conf` http {} section.

Perfect, that was it, thanks a lot. TL;DR I can't reproduce, things look fine on my machine. Can you push your Vike project somewhere? Do things work for you with the react-full-v1 example? (see below)

#### Details

I added the `server {}` snippet at the end of the `http {}` section of my nginx config.
I added `127.0.0.1 sitename` to my `/etc/hosts`
I checked out [the react-full-v1 example](https://github.com/vikejs/vike/tree/main/examples/react-full-v1) which uses client routing and modified it to listen to a different port:

```diff
--- a/examples/react-full-v1/package.json
+++ b/examples/react-full-v1/package.json
@@ -1,6 +1,6 @@
 {
   "scripts": {
-    "dev": "vite",
+    "dev": "vite --port 5173",
     "build": "vite build",
     "preview": "vite build && vite preview"
   },
```

I ran `pnpm run dev` and opened http://sitename in my firefox browser, and navigated to all the pages. Everything flowed well.

Now your Vike project is probably quite different since you're using Svelte (can you push it somewhere?), and also probably we're still having different nginx configs. Does the react-full-v1 example work for you or also not?

**Reply by igolkotek** [reply]
@AurelienLourot ,

I just tried bare projects generated by pnpm create vike (both vue & react).
OK, I'll try the react-full-v1 example )

**Reply by igolkotek** [reply]
@AurelienLourot ,

I've tested the `react-full-v1` example, and it's OK.
The difference between this example and the `vike-react template` is an express server with `renderPage`. 
And I need it to have a customised pageContext.

**Reply by igolkotek** [reply]
I added server code from vike-react template and changed it from using body to pipe according to the documentation.
```
    const pageContext = await renderPage(pageContextInit)
    const { httpResponse } = pageContext
    if (!httpResponse) return next()
    httpResponse.pipe(res)
```

And it works like a charm.. %-)

Is the reason in using stream?

**Reply by lourot** [reply]
Ah indeed, I can reproduce with the `react-streaming-v1` example (`PORT=5173 pnpm run dev`). I suspect nginx needs to be configured a bit differently to play nicely with streaming HTML. (Maybe the communication then looks a bit different at HTTP or TCP level? Longer packets?) Do you need streaming?

**Reply by igolkotek** [reply]
Why?
There are no problems with streaming while testing, they're with regular http. Nginx default settings have no issues with http at all - proved by years of usage.

**Reply by lourot** [reply]
It would be important to get to the bottom of it but if it's not a blocker for you (let me know if it is), let's keep it open for now as I'm working on other issues.

If you feel like investigating this yourself it would help us a lot too. Any kinds of contribution are always welcome. Also did you consider [sponsoring](https://github.com/sponsors/vikejs/)? It makes a huge difference for us and you get a priority bump on your issues.

**Reply by igolkotek** [reply]
OK, @AurelienLourot , thank you for your help.
The reason of the issue still not found, but we got closer.

I do not need a stream, so will try to isolate the reason. 
Also would take into account a sponsorship )
